cli/npm/resolution/graph.rs

```diff
@@ -690,496 +962,338 @@ impl<'a, TNpmRegistryApi: NpmRegistryApi>
   fn resolve_peer_dep(
     &mut self,
     specifier: &str,
-    parent_id: &NpmPackageNodeId,
     peer_dep: &NpmDependencyEntry,
     peer_package_info: &NpmPackageInfo,
-    visited_ancestor_versions: &Arc<VisitedVersionsPath>,
-    existing_dep_id: Option<&NpmPackageNodeId>,
-  ) -> Result<Option<NpmPackageNodeId>, AnyError> {
-    fn find_matching_child<'a>(
-      peer_dep: &NpmDependencyEntry,
-      peer_package_info: &NpmPackageInfo,
-      children: impl Iterator<Item = &'a NpmPackageNodeId>,
-    ) -> Result<Option<NpmPackageNodeId>, AnyError> {
-      for child_id in children {
-        if child_id.name == peer_dep.name
-          && version_req_satisfies(
-            &peer_dep.version_req,
-            &child_id.version,
-            peer_package_info,
-            None,
-          )?
-        {
-          return Ok(Some(child_id.clone()));
-        }
+    ancestor_path: &Arc<GraphPath>,
+  ) -> Result<Option<NodeId>, AnyError> {
+    debug_assert!(matches!(
+      peer_dep.kind,
+      NpmDependencyEntryKind::Peer | NpmDependencyEntryKind::OptionalPeer
+    ));
+
+    // Use this to detect cycles. This is just in case and probably won't happen
+    let mut up_path = GraphPath::new(ancestor_path.node_id());
+    let mut path = vec![ancestor_path];
+
+    // the current dependency might have had the peer dependency
+    // in another bare specifier slot... if so resolve it to that
+    {
+      let maybe_peer_dep = self.find_peer_dep_in_node(
+        ancestor_path,
+        peer_dep,
+        peer_package_info,
+      )?;
+
+      if let Some((peer_parent, peer_dep_id)) = maybe_peer_dep {
+        // this will always have an ancestor because we're not at the root
+        self.set_new_peer_dep(peer_parent, path, specifier, peer_dep_id);
+        return Ok(Some(peer_dep_id));
       }
-      Ok(None)
     }
 
     // Peer dependencies are resolved based on its ancestors' siblings.
     // If not found, then it resolves based on the version requirement if non-optional.
-    let mut pending_ancestors = VecDeque::new(); // go up the tree by depth
-    let path = GraphSpecifierPath::new(specifier.to_string());
-    let visited_versions = VisitedVersionsPath::new(parent_id);
-
-    // skip over the current node
-    for (specifier, grand_parents) in
-      self.graph.borrow_node(parent_id).parents.clone()
-    {
-      let path = path.with_specifier(specifier);
-      for grand_parent in grand_parents {
```

Comment:
This was previously so wrong it's embarrassing, but it seems like it's worked out ok. The issue here is that we can't look at all of a node's parents... we have to just look at the current graph path's ancestors.